compiler/fir/evaluate/src/org/jetbrains/kotlin/fir/evaluate/CompileTimeConstantEvaluator.kt

```diff
@@ -0,0 +1,225 @@
+        return try {
+            val r = f(typedValue, other.typedValue)
+            r.toFirConstKind()!!.toConstExpression(source, r)
+        } catch (e: Exception) {
+            null
+        }
+    }
+
+    ////// KINDS
+
+    private fun FirTypeRef.toFirConstKind(): FirConstKind<out Number>? =
+        when (this) {
+            !is FirResolvedTypeRef -> null
+            !is FirImplicitBuiltinTypeRef -> type.toFirConstKind()
+            is FirImplicitIntTypeRef -> FirConstKind.Int
+            else -> null
+        }
+
+    private fun ConeKotlinType.toFirConstKind(): FirConstKind<out Number>? =
+        when (this) {
+            is ConeKotlinErrorType -> null
+            is ConeLookupTagBasedType -> {
+                when (lookupTag.name.asString()) {
+                    "Byte" -> FirConstKind.Byte
+                    "Double" -> FirConstKind.Double
+                    "Float" -> FirConstKind.Float
+                    "Int" -> FirConstKind.Int
+                    "Long" -> FirConstKind.Long
+                    "Short" -> FirConstKind.Short
+                    else -> null
+                }
+            }
+            is ConeFlexibleType -> upperBound.toFirConstKind()
+            is ConeCapturedType -> lowerType?.toFirConstKind() ?: constructor.supertypes!!.first().toFirConstKind()
+            is ConeDefinitelyNotNullType -> original.toFirConstKind()
+            is ConeIntersectionType -> intersectedTypes.first().toFirConstKind()
+            is ConeStubType -> null
+            is ConeIntegerLiteralType -> null
+        }
+
+    private fun <T : Number> T.toFirConstKind(): FirConstKind<out Number>? =
+        when (this) {
+            is Byte -> FirConstKind.Byte
+            is Double -> FirConstKind.Double
+            is Float -> FirConstKind.Float
+            is Int -> FirConstKind.Int
+            is Long -> FirConstKind.Long
+            is Short -> FirConstKind.Short
+            else -> null
+        }
+
+    private fun FirConstKind<*>.convert(value: Number?): Number? {
+        if (value == null) {
+            return null
+        }
+        return when {
+            this == FirConstKind.Byte -> value.toByte()
+            this == FirConstKind.Double -> value.toDouble()
+            this == FirConstKind.Float -> value.toFloat()
+            this == FirConstKind.Int -> value.toInt()
+            this == FirConstKind.Long -> value.toLong()
+            this == FirConstKind.Short -> value.toShort()
+            else -> null
+        }
+    }
+
+    private fun <T : Number> FirConstKind<T>.toConstExpression(source: FirSourceElement?, value: Number): FirConstExpression<T> =
+        @Suppress("UNCHECKED_CAST")
+        buildConstExpression(source, this, value as T)
+
+    // TODO: extension functions, e.g., Byte.and
+    // TODO: toChar?, toString?
+    companion object {
+        private val unaryOperatorNames: Set<String> = setOf(
```

Comment:
Just an idea: Instead of hard-coding (prone to being out-of-sync), you could put both of these as properties in `FirOperationsMapGenerated`, something like:
```kt
internal val unaryOperationsMap = unaryOperations.map { it.opName }.toSet()
```

Reply:
That's good idea, thanks! Changes added to the fifth commit..;;

Reply:
Oops I actually meant `unaryOperations.keys.map { ... }` but your code works the same. 🙂